src/js/chat.js

```diff
@@ -0,0 +1,342 @@
+    let index = Math.abs(hash % COLORS.length);
+    return COLORS[index];
+  }
+
+  bindEvents() {
+    var _this = this;
+    // Select message input when clicking message body, unless selecting text
+    this.messages.on('click', () => {
+      if (!this.getSelectedText()) {
+        this.inputMessage.focus();
+      }
+    });
+
+    this.inputMessage.on('input propertychange paste change', function() {
+      _this.updateTyping();
+      let message = $(this).val().trim();
+      if (message.length) {
+        $('#send-message-btn').addClass('active');
+      } else {
+        $('#send-message-btn').removeClass('active');
+      }
+    });
+  }
+
+  // Updates the typing event
+  updateTyping() {
+    if (this.darkwire.connected) {
+      if (!this.typing) {
+        this.typing = true;
+        this.socket.emit('typing');
+      }
+      this.lastTypingTime = (new Date()).getTime();
+
+      setTimeout(() => {
+        let typingTimer = (new Date()).getTime();
+        let timeDiff = typingTimer - this.lastTypingTime;
+        if (timeDiff >= this.TYPING_TIMER_LENGTH && this.typing) {
+          this.socket.emit('stop typing');
+          this.typing = false;
+        }
+      }, this.TYPING_TIMER_LENGTH);
+    }
+  }
+
+  addChatTyping(data) {
+    data.typing = true;
+    data.message = 'is typing';
+    this.addChatMessage(data);
+  }
+
+  getSelectedText() {
+    let text = '';
+    if (typeof window.getSelection != 'undefined') {
+      text = window.getSelection().toString();
+    } else if (typeof document.selection != 'undefined' && document.selection.type == 'Text') {
+      text = document.selection.createRange().text;
+    }
+    return text;
+  }
+
+  getTypingMessages(data) {
+    return $('.typing.message').filter(function(i) {
+      return $(this).data('username') === data.username;
+    });
+  }
+
+  removeChatTyping(data) {
+    this.getTypingMessages(data).fadeOut(function() {
+      $(this).remove();
+    });
+  }
+
+  slashCommands(trigger) {
+    let validCommands = [];
+    let expectedParams = 0;
+    const triggerCommands = [{
+      command: 'nick',
+      description: 'Changes nickname.',
+      paramaters: ['{username}'],
+      multiple: false,
+      usage: '/nick {username}',
+      action: () => {
+        let newUsername = trigger.params[0] || false;
+
+        if (newUsername > 16) {
+          return this.log('Username cannot be greater than 16 characters.', {error: true});
+        }
+
+        // Remove things that arent digits or chars
+        newUsername = newUsername.replace(/[^A-Za-z0-9]/g, '-');
+
+        if (!newUsername.match(/^[A-Z0-9]/i)) {
+          return this.log('Username must start with a letter or number.', {error: true});
+        }
+
+        this.darkwire.updateUsername(newUsername).then((socketData) => {
+          let modifiedSocketData = {
+            username: window.username,
+            newUsername: socketData.username,
+            publicKey: socketData.publicKey
```

Comment:
Users will already have everyone else's public key, so no need to send it. We can have the recipient find the public key using the sender's ID. Check out how message sending works for an example.